### PR TITLE
feat(atomic): new popover facet component

### PR DESCRIPTION
**Pull request by mbenkirane-coveo**
https://coveord.atlassian.net/browse/KIT-1840

### CODE REVIEW

**Comment by mbenkirane-coveo** [comment]
Will merge with master after getting a couple reviews !

File: packages/atomic/src/components/common/atomic-focus-trap.tsx [hunk 1]
```diff
@@ -1,34 +1,11 @@
...
-  }
-  const parent = getParent(element);
-  return parent === null ? false : contains(ancestor, parent);
-}
+import {
+  defer,
+  getFocusedElement,
+  getParent,
+  isAncestorOf,
+} from '../../utils/utils';
```
Comment (btaillon) [inline packages/atomic/src/components/common/atomic-focus-trap.tsx:8]: for anyone wondering: this was copied from the changes I need to make for the search box in order to implement `atomic-focus-detector`. I think it's fine for this change to stay in this PR.

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  @Element() host!: HTMLElement;
+  private facetElement?: HTMLElement;
+
+  public initialize() {
+    this.searchStatus = buildSearchStatus(this.bindings.engine);
+  }
+
+  popoverOpened() {
+    if (!this.isMenuVisible) {
+      const popoverOpened = new CustomEvent('popoverOpened', {
```
Comment (btaillon) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:56]: After all, I think that this event should be called something like `clearModals`, since I can imagine actions unrelated to opening a popover wanting to close the popover.

Comment (btaillon) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:56]: I would also make the type of the event more explicit by adding a generic type to represent details and naming the facetId prop.

E.g.:
```ts
interface ClearModalsEventHandler {
  sourcePopoverFacetId?: string;
}
...
const popoverOpened = new CustomEvent<ClearModalsEventHandler>(
  'popoverOpened',
  {
    details: { sourcePopoverFacetId: this.facetId },
  }
);
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+          {this.facetLabel}
+        </span>
+        <atomic-icon
+          part="arrow-icon"
+          class={`w-2 ml-2 ${
+            this.isMenuVisible
+              ? 'rotate-180'
+              : 'group-hover:text-primary-light group-focus:text-primary'
+          } `}
+          icon={ArrowBottomIcon}
```
Comment (btaillon) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:102]: I would add `aria-hidden="true` to the icon to remove it from screen readers' accessibility DOM, since it's adequately described by the button.

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    return (
+      <Button
+        style="square-neutral"
+        onClick={() => this.popoverOpened()}
+        part="value-button"
+        class={`value-button rounded flex box-border h-full items-center mr-1.5 p-2.5 group ${
+          this.isMenuVisible
+            ? 'selected border-primary shadow-outer-light-blue text-primary'
+            : 'hover:border-primary-light focus-visible:border-primary-light '
+        }`}
```
[the remark this answers is not in the document]
Comment (btaillon) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:82]: I would also add an [`aria-label`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) to this button, which reads to screen reader both the facet label and something that indicates that this is an expandable facet. You could copy the label of headers of normal facets.

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    if (this.facetElement) {
+      return;
+    }
+
+    const facet = this.bindings.store.get(event.detail.facetType)[
+      event.detail.facetId
+    ];
+
+    if (!facet) {
+      // TODO: add error msg
```
Comment (btaillon) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:119]: Is this meant to be done in this PR or another PR?

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+  renderValueButton() {
+    return (
+      <Button
+        style="square-neutral"
+        onClick={() => this.popoverOpened()}
+        part="value-button"
+        class={`value-button rounded flex box-border h-full items-center mr-1.5 p-2.5 group ${
+          this.isMenuVisible
+            ? 'selected border-primary shadow-outer-light-blue text-primary'
```
Comment (btaillon) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:80]: Instead of `shadow-outer-light-blue`, I think what you may be looking for is `ring-2 ring-ring-primary`.

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+} from '../../../../utils/initialization-utils';
+import {Button} from '../../../common/button';
+import {Hidden} from '../../../common/hidden';
+import {Bindings} from '../../atomic-search-interface/atomic-search-interface';
+import ArrowBottomIcon from 'coveo-styleguide/resources/icons/svg/arrow-bottom-rounded.svg';
+import {InitPopoverEventPayload} from './popover-event';
+
+/**
+ * @internal
+ * The `atomic-popover` displays a popover menu view of any facet
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:20]:
```suggestion
 * The `atomic-popover` component displays any facet as a popover menu.
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+ * @part value-button - The button used to display the popover menu
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:21]:
```suggestion
 * @part value-button - The button to click to display the popover menu.
```
Also, I'm curious: why not `button` rather than`value-button`?
And I'm also wondering the same thing about `label` vs `value-label` 🤔

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+ * @part value-label - The popover value label
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:22]:
```suggestion
 * @part value-label - The popover button label.
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+ * @part arrow-icon - The arrow icon
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:23]:
```suggestion
 * @part arrow-icon - The arrow icon to display or hide the popover menu.
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+ * @part placeholder - The placeholder displayed when the facet isn't ready
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:24]:
```suggestion
 * @part placeholder - The placeholder displayed when the facet is loading.
```
(assuming I understand)

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+ * @part popover-wrapper - The wrapper that contains the value-button and the slot-wrapper
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:25]:
```suggestion
 * @part popover-wrapper - The wrapper that contains the 'value-button' and the 'slot-wrapper'.
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+ * @part slot-wrapper - The wrapper that contains the facet/slot
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:26]:
```suggestion
 * @part slot-wrapper - The wrapper that contains the 'facet' or 'slot'.
```
What's a slot in this context? First time it's mentioned in this file 🤔

File: packages/atomic/src/pages/index.html [hunk 14]
```diff
@@ -201,6 +205,17 @@
               number-of-values="20"
             ></atomic-segmented-facet>
           </atomic-segmented-facet-scrollable>
+          <atomic-layout-section class="flex" section="popovers">
+            <atomic-popover>
+              <atomic-facet field="objecttype" number-of-values="8"></atomic-facet>
```
Comment (ThibodeauJF) [inline packages/atomic/src/pages/index.html:210]: Need to add ids `facet-id="something"` to all these facets & the segmented one since the field are already used for other facets

File: packages/atomic/src/pages/index.html [hunk 15]
```diff
@@ -128,6 +128,10 @@
       atomic-search-layout {
         row-gap: var(--atomic-layout-spacing-y);
       }
+
+      atomic-layout-section[section='popovers'] {
+        display: flex;
+      }
```
Comment (ThibodeauJF) [inline packages/atomic/src/pages/index.html:134]: should be moved to packages/atomic/src/components/search/atomic-layout/atomic-search-layout.pcss

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+        message: 'Cannot have more than one child inside a set of popover tags',
+      };
+      return (
+        <atomic-component-error
+          element={this.host}
+          error={this.error}
+        ></atomic-component-error>
+      );
+    }
+
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:157]: When error is defined it should render atomic-component-error automatically.
You can move this check higher up in the process, like the connected callback or initialize method

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx (hunk 13 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:26]: A slot is a placeholder that can be filled with another component

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx (hunk 11 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:24]: You're right !

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx (hunk 8 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:21]: I should probably call them popover-button and label. Maybe that would make it more clear?

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx (hunk 5 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:119]: I'll add it on this PR

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx (hunk 2 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:56]: I'm not sure I would go with the term modal. I would use dropdown instead.

File: packages/atomic/src/pages/index.html (hunk 15 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/pages/index.html:134]: But popovers is an arbitrary section name. Shouldn't clients be able to name their section whatever name they'd like?

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+import ArrowBottomIcon from 'coveo-styleguide/resources/icons/svg/arrow-bottom-rounded.svg';
+import {
+  InitPopoverEventPayload,
+  ClearPopoversEventPayload,
+} from './popover-event';
+
+/**
+ * @internal
+ * The `atomic-popover` component displays any facet as a popover menu.
+ * @part label-button - The button used to display the popover menu
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:24]: I think that your "popover-button" was better 👍 
```suggestion
 * @part popover-button - The button to click to display or hide the popover menu.
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+ * @part label - The popover button label.
+ * @part arrow-icon - The arrow icon to display or hide the popover menu.
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:26]:
```suggestion
 * @part arrow-icon - The arrow icon on the button to display or hide the popover menu.
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+ * @part placeholder - The placeholder displayed when the facet is loading.
+ * @part popover-wrapper - The wrapper that contains the 'value-button' and the 'slot-wrapper'.
```
Comment (jpmarceau) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:28]:
```suggestion
 * @part popover-wrapper - The wrapper that contains the 'popover-button' and the 'slot-wrapper'.
```

**Review by jpmarceau — APPROVED** [review]

**Review by btaillon — APPROVED** [review]

File: packages/atomic/src/components/search/facets/atomic-facet/atomic-facet.tsx [hunk 20]
```diff
@@ -245,6 +257,13 @@ export class AtomicFacet implements InitializableComponent, BaseFacet<Facet> {
     });
   }
 
+  public componentDidRender() {
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-facet/atomic-facet.tsx:260]: Is it working if you put this emitter in the initialize method? The earlier the better

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.pcss [hunk 21]
```diff
@@ -0,0 +1,6 @@
+@import '../../../../global/global.pcss';
+@import '../facet-common.pcss';
+
+.popover-wrapper .slot-wrapper.selected {
+  display: block;
+}
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.pcss:6]: Could you remove this & just put the `block` tailwind class instead of selected?

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+ * @internal
+ * The `atomic-popover` component displays any facet as a popover menu.
+ * @part popover-button - The button to click to display or hide the popover menu.
+ * @part label - The popover button label.
+ * @part arrow-icon - The arrow icon on the button to display or hide the popover menu.
+ * @part placeholder - The placeholder displayed when the facet is loading.
+ * @part popover-wrapper - The wrapper that contains the 'popover-button' and the 'slot-wrapper'.
+ * @part slot-wrapper - The wrapper that contains the 'facet' or 'slot'.
+ */
+
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:31]:
```suggestion
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+      this.error = {
+        name: 'No child in popover',
+        message: 'One child is required inside a set of popover tags',
+      };
+    } else if (this.host.children.length > 1) {
+      this.error = {
+        name: 'Too many children in popover',
+        message: 'Cannot have more than one child inside a set of popover tags',
+      };
+    }
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:66]: we generally prefer early returns rather than else ifs
```suggestion
    if (this.host.children.length === 0) {
      this.error = {
        name: 'No child in popover',
        message: 'One child is required inside a set of popover tags',
      };
      return;
    } 
    
    if (this.host.children.length > 1) {
      this.error = {
        name: 'Too many children in popover',
        message: 'Cannot have more than one child inside a set of popover tags',
      };
    }
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+      <Button
+        style="square-neutral"
+        onClick={() => this.popoverOpened()}
+        part="label-button"
+        ariaExpanded={`${this.isMenuVisible}`}
+        ariaLabel={`Pop-up filter on ${this.facetLabel} facet`}
+        ariaControls={this.popoverId}
+        class={`rounded flex box-border h-full items-center mr-1.5 p-2.5 group ${
+          this.isMenuVisible
+            ? 'selected border-primary ring ring-ring-primary-light text-primary'
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:105]: related to css comment
```suggestion
            ? 'block border-primary ring ring-ring-primary-light text-primary'
```

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+              : 'group-hover:text-primary-light group-focus:text-primary'
+          } `}
+          icon={ArrowBottomIcon}
+        ></atomic-icon>
+      </Button>
+    );
+  }
+
+  @Listen('facetInitialized')
+  injectPopoverClass(event: CustomEvent<InitPopoverEventPayload>) {
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:135]: Perhaps a better name would be linkFacet or something similar, because there is more to it than injecting the class

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+    return (
+      <div
+        id={this.popoverId}
+        part="popover-wrapper"
+        class="popover-wrapper relative"
+      >
+        {this.renderValueButton()}
+        <div
+          part="slot-wrapper"
+          class={`slot-wrapper absolute pt-2 z-10 hidden ${
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:187]:
```suggestion
          class={`absolute pt-2 z-10 hidden ${
```

File: packages/atomic/src/pages/index.html (hunk 15 above)
Comment (ThibodeauJF) [reply packages/atomic/src/pages/index.html:134]: We offer a basic search layout with predefined names & styling. https://docs.coveo.com/en/atomic/latest/usage/layouts/

Customer have the choice of using it or not, but we should provide a section for these components. Could be a section for all the horizontal facets, not just he popovers.

I'd name the section `horizontal-filters` & move all the horizontal facets inside, similar to the mockups:
Screen Shot 2022-08-04 at 12 20 48 PM

File: packages/atomic/src/themes/coveo.css [hunk 27]
```diff
@@ -5,6 +5,7 @@
   --atomic-primary-dark: #1a50ad;
   --atomic-on-primary: #ffffff;
   --atomic-ring-primary: rgba(19, 114, 236, 0.5);
+  --atomic-ring-primary-light: rgba(57, 159, 254, 0.3);
```
Comment (ThibodeauJF) [inline packages/atomic/src/themes/coveo.css:8]: This extra variable does not seem worth it, I'd prefer reusing `--atomic-ring-primary` & keep the number of theme variables to a minimum. Users will have more control if they chose

e.g.

```
atomic-popover {
  --atomic-ring-primary: rgba(57, 159, 254, 0.3)
}
```

File: packages/atomic/tailwind.config.js [hunk 28]
```diff
@@ -33,7 +34,8 @@ module.exports = {
         xl: 'var(--atomic-border-radius-xl)',
       },
       boxShadow: {
-        'inner-primary': 'inset 0 0 0 1px var(--atomic-primary)'
+        'inner-primary': 'inset 0 0 0 1px var(--atomic-primary)',
+        'outer-light-blue': '0px 0px 0px 3px rgba(57, 159, 254, 0.3)'
```
Comment (ThibodeauJF) [inline packages/atomic/tailwind.config.js:38]:
```suggestion
```

File: packages/samples/atomic-react/src/style.css [hunk 29]
```diff
@@ -20,3 +20,7 @@ header ul {
   font-weight: var(--atomic-font-bold);
   font-family: var(--atomic-font-family);
 }
+
+atomic-popover atomic-facet::part(label-button) {
+  display: none;
+}
```
Comment (ThibodeauJF) [inline packages/samples/atomic-react/src/style.css:26]: Why is this needed?

File: packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx (hunk 24 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-popover/atomic-popover.tsx:105]: actually we can remove it the selected tag

File: packages/atomic/src/pages/index.html (hunk 15 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/pages/index.html:134]: I see !

File: packages/atomic/src/components/search/facets/atomic-facet/atomic-facet.tsx (hunk 20 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/components/search/facets/atomic-facet/atomic-facet.tsx:260]: Yup, should work !

File: packages/atomic/src/themes/coveo.css (hunk 27 above)
Comment (mbenkirane-coveo) [reply packages/atomic/src/themes/coveo.css:8]: They'd have to add it on their own on their custom css file though?
